### ISSUE TITLE
'bool' object is not iterable 

### ISSUE THREAD

**Issue by kylekam** [ultralytics/ultralytics#7443]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Val

### Bug

I'm trying to train yolov8m on a custom dataset by specifying a yaml file with the proper fields. I saw that the "val" field is used both to specify the path to validation files AND for a boolean to enable or disable val/test during training. I believe this is why the "bool object is not iterable" error is occuring. The error occurs on this line https://github.com/ultralytics/ultralytics/blob/c26d4a222fa812c77fcba3cf904165f97f559a0f/ultralytics/data/utils.py#L299C53-L299C54.

My default.yaml file looks something like the following: 

task: detect  # inference task, i.e. detect, segment, classify
mode: train  # YOLO mode, i.e. train, val, predict, export

path: /home/datascience/tooltracking_data/
train: ./train
test: ./test
val: ./val

nc: 1

names: ['tool1', 'tool2']

model:  # path to model file, i.e. yolov8n.pt, yolov8n.yaml
data:  # path to data file, i.e. i.e. coco128.yaml
epochs: 400  # number of epochs to train for
patience: 75  # epochs to wait for no observable improvement for early stopping of training
batch: 32  # number of images per batch (-1 for AutoBatch)
imgsz: 640  # size of input images as integer or w,h
save: True  # save train checkpoints and predict results
cache: False  # True/ram, disk or False. Use cache for data loading
device: 0 # device to run on, i.e. cuda device=0 or device=0,1,2,3 or device=cpu
workers: 8  # number of worker threads for data loading (per RANK if DDP)
project:  # project name
name:  # experiment name
exist_ok: False  # whether to overwrite existing experiment
pretrained: False  # whether to use a pretrained model
optimizer: SGD  # optimizer to use, choices=['SGD', 'Adam', 'AdamW', 'RMSProp']
verbose: True  # whether to print verbose output
seed: 0  # random seed for reproducibility
deterministic: True  # whether to enable deterministic mode
single_cls: False  # train multi-class data as single-class
image_weights: False  # use weighted image selection for training
rect: False  # support rectangular training
cos_lr: False  # use cosine learning rate scheduler
close_mosaic: 10  # disable mosaic augmentation for final 10 epochs
resume: False  # resume training from last checkpoint

overlap_mask: True  # masks should overlap during training (segment train only)
mask_ratio: 4  # mask downsample ratio (segment train only)

dropout: 0.0  # use dropout regularization (classify train only)

val: True  # validate/test during training
save_json: False  # save results to JSON file
save_hybrid: False  # save hybrid version of labels (labels + additional predictions)
conf:  # object confidence threshold for detection (default 0.25 predict, 0.001 val)
iou: 0.7  # intersection over union (IoU) threshold for NMS
max_det: 300  # maximum number of detections per image
half: False  # use half precision (FP16)
dnn: False  # use OpenCV DNN for ONNX inference
plots: True  # save plots during train/val

source:  # source directory for images or videos
show: False  # show results if possible
save_txt: False  # save results as .txt file
save_conf: False  # save results with confidence scores
save_crop: False  # save cropped images with results
hide_labels: False  # hide labels
hide_conf: False  # hide confidence scores
vid_stride: 1  # video frame-rate stride
line_thickness: 3  # bounding box thickness (pixels)
visualize: False  # visualize model features
augment: False  # apply image augmentation to prediction sources
agnostic_nms: False  # class-agnostic NMS
classes:  # filter results by class, i.e. class=0, or class=[0,2,3]
retina_masks: False  # use high-resolution segmentation masks
boxes: True # Show boxes in segmentation predictions

format: torchscript  # format to export to
keras: False  # use Keras
optimize: False  # TorchScript: optimize for mobile
int8: False  # CoreML/TF INT8 quantization
dynamic: False  # ONNX/TF/TensorRT: dynamic axes
simplify: False  # ONNX: simplify model
opset:  # ONNX: opset version (optional)
workspace: 4  # TensorRT: workspace size (GB)
nms: False  # CoreML: add NMS

lr0: 0.01  # initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01  # final learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 7.5  # box loss gain
cls: 0.5  # cls loss gain (scale with pixels)
dfl: 1.5  # dfl loss gain
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
label_smoothing: 0.15  # label smoothing (fraction)
nbs: 64  # nominal batch size
hsv_h: 0.15  # image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # image HSV-Value augmentation (fraction)
degrees: 180.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.5  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.5  # image flip up-down (probability)
fliplr: 0.5  # image flip left-right (probability)
mosaic: 0.5  # image mosaic (probability)
mixup: 0.0  # image mixup (probability)
copy_paste: 0.0  # segment copy-paste (probability)

cfg:  # for overriding defaults.yaml

v5loader: False  # use legacy YOLOv5 dataloader

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@kylekam hello! It seems like there's a conflict in your `default.yaml` due to the `val` key being used twice with different data types: once as a path to the validation dataset and once as a boolean to enable or disable validation during training. To resolve this, you should only use the `val` key once and assign it the path to your validation dataset. The boolean to enable or disable validation during training is controlled by a different parameter, which is not specified by the `val` key.

Please ensure that each key in your YAML configuration file is unique and correctly assigned to prevent such conflicts. If you need further guidance on configuring your YAML file, you can refer to the documentation at https://docs.ultralytics.com for detailed instructions on setting up your training environment.

If after making these adjustments you still encounter issues, feel free to reach out with the updated configuration, and we'll be happy to assist you further. Good luck with your training! 🚀

**Comment by kylekam** [comment]
I acknowledge that 'val' should be used to specify the path as done here: https://github.com/ultralytics/ultralytics/issues/2650. But then what should the correct name be for the field that enable or disables the validation during training? The example provided does not solve this: https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml

**Comment by glenn-jocher** [comment]
@kylekam apologies for the confusion. The `val` key in your YAML should indeed specify the path to your validation dataset. To enable or disable validation during training, you should use the `val` key under the `train` section as a boolean. If this key is not present, validation is enabled by default. If you wish to disable it, you can set `val: False` within the `train` section of your configuration. Please refer to the documentation for further details on configuring the training process. 🛠️

**Comment by kylekam** [comment]
My mistake, I combined the two YAML files when they should've been separate. https://github.com/ultralytics/ultralytics/issues/2650 describes the datset.yaml passed in during training as a parameter and https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/default.yaml describes the default paramters and hyperparameters for training.

**Comment by glenn-jocher** [comment]
No worries, @kylekam! You've got it right now. The `dataset.yaml` is for specifying your dataset paths and related settings, while `default.yaml` contains the default training parameters and hyperparameters. Keep them separate, and you should be good to go. Happy training! 🚀
